### PR TITLE
[ci] Change macOS GHA runner to `macos-latest-xl-arm64`

### PR DESCRIPTION
**Pull request by reedloden**
Swap from Intel-based `macos-12-xl` to Apple silicon-based `macos-latest-xl-arm64` runner.

Refs https://github.com/gravitational/teleport/issues/28725

### CODE REVIEW

File: .github/workflows/build-macos.yaml [hunk 1]
```diff
@@ -2,6 +2,9 @@ name: Build on Mac OS
 run-name: Build on Mac OS
 
 on:
+  pull_request:
+    paths:
+      - '.github/workflows/build-macos.yaml'
```
Comment (r0mant) [inline .github/workflows/build-macos.yaml:7]: @reedloden I've added this temporarily here to see what happens with the macOS build.

Comment (r0mant) [reply .github/workflows/build-macos.yaml:7]: And it fails with a permission issue: https://github.com/gravitational/teleport/actions/runs/5469360603/jobs/9958206534?pr=28741.

@camscale Could you take a look?

Comment (camscale) [reply .github/workflows/build-macos.yaml:7]: Ok. pushing a new commit to see what's going on

**Comment by camscale** [comment]
On the new ARM64 runners, the permissions of the cache directory are:
```
drwxr-xr-x  3 root  staff  96 Jun 14 04:04 /Users/runner/.cache
```

On the old AMD64 runners they are:
```
drwxr-xr-x  3 runner  staff  96 Jun 14 04:04 /Users/runner/.cache
```

The ownership is wrong. It's possible that the rest of the `/Users/runner` directory has the wrong ownership too. This is an easy fix with a `chown` step at the start, but we shouldn't have to do this. Can we give this feedback to GitHub?

**Comment by reedloden** [comment]
> Can we give this feedback to GitHub?

Submitted https://support.github.com/ticket/enterprise/9266/2238925

**Comment by camscale** [comment]
> Submitted https://support.github.com/ticket/enterprise/9266/2238925

Looks like the `.config` directory is also root-owned. So it is `.cache` and `.config` that are root owned. Everything else is owned by the runner user.

**Review by r0mant — APPROVED** [review]
@camscale Looks like this is working now? macOS build takes 5 minutes, almost 2x improvements, nice.

**Review by jentfoo — APPROVED** [review]

**Review by wadells — APPROVED** [review]
If we want to be extra sure this is safe, we should do the files in `teleport.e`:

https://github.com/gravitational/teleport.e/blob/master/.github/workflows/build-mac.yaml
https://github.com/gravitational/teleport.e/blob/master/.github/workflows/build-mac-universal.yaml

Until those get the change, we'll be in a split brain mode, where we do CI on ARM, but release off Intel.

**Comment by camscale** [comment]
> If we want to be extra sure this is safe, we should do the files in teleport.e:

I have some changes ready to go for that already but:
1. Wanted to wait to see how quickly github respond and fix the permission issue or whether I should put that step in to fix it (I currently have the step in there)
2. Hit the rate limit issue on drone so couldn't test it to be sure.

I'll give it a run today now that it looks like rate limiting might have been alleviated for now and check that it works for the release build and if so, raise a PR.

**Comment by wadells** [comment]
> Wanted to wait to see how quickly github respond and fix the permission issue 

Alas, no response to the issue yet.

**Comment by camscale** [comment]
The related PR for the release workflows is https://github.com/gravitational/teleport.e/pull/1790

**Comment by camscale** [comment]
@reedloden I'm going to merge this - the equivalent changes have been made in teleport.e. This workflow does not need the additional change of installing pkg-config as this workflow does not build fido2. It looks like the runner image still has not been updated (runner image is 20230611.2 according to the [readme](https://github.com/actions/runner-images/blob/main/images/macos/macos-13-Readme.md)).

**Pull request merged**